### ISSUE TITLE
/LOCAL refinement injected by function constructors is malformed

### ISSUE THREAD

**Issue by 9214** [red/red#4537]
**Describe the bug**
`has` and `function` inject `/local` refinement into constructed function's spec. Such refinement is an `any-word!` value, and, as such, must be bound to a context, which turns out not to be true on practice.

**To reproduce**
Any of the examples below:
```red
get first spec-of has [foo][]
get first spec-of function [][foo:]
```
Results in:
```red
*** Runtime Error 1: access violation
*** in file: .../red/runtime/datatypes/context.reds
*** at line: 325
***
***   stack: red/_context/get 02C8A414h
***   stack: red/natives/get* false -1 -1
***   stack: red/interpreter/eval-arguments 02C8A404h 031FFCC8h 031FFCC8h 00000000h 00000000h
***   stack: red/interpreter/eval-code 02C8A404h 031FFC98h 031FFCC8h false 00000000h 00000000h 02E50C38h
***   stack: red/interpreter/eval-expression 031FFC98h 031FFCC8h false false false
***   stack: red/interpreter/eval 02C8A3F4h true
***   stack: red/natives/try* true 0
***   stack: do-file
***   stack: red/natives/do* false -1 -1 -1
***   stack: red/interpreter/eval-arguments 02C8A334h 031FF298h 031FF298h 00000000h 00000000h
***   stack: red/interpreter/eval-code 02C8A334h 031FF288h 031FF298h false 00000000h 00000000h 02E50BA8h
***   stack: red/interpreter/eval-expression 031FF288h 031FF298h false false false
***   stack: red/interpreter/eval 02C8A314h true
***   stack: red/natives/catch* true 1
***   stack: ctx||564~try-do 026E1D2Ch
***   stack: ctx||564~do-command 026E1D2Ch
***   stack: ctx||564~eval-command 026E1D2Ch
***   stack: ctx||564~run 026E1D2Ch
***   stack: ctx||564~launch 026E1D2Ch
***   stack: ctx||598~launch 026E14B4h
```

Because context node is `null`.

**Expected behavior**
I expect function's refinements (including `/local`) to be bound to either global context or function's context.

**Platform version**
4d864b1d

**Comment by greggirwin** [comment]
This is interesting, and a deep, fundamental question. _Do_ words have to be bound to a context, or can words _not_ be bound, meaning they are only ever interpretable as literal data? `/Local` is certainly more in the function's context than global, but also in a sort of limbo as it's just a marker. The next question is what `get /local` should return? What do you expect in this case?

**Comment by greggirwin** [comment]
I certainly agree that it shouldn't crash though.

**Comment by meijeru** [comment]
Compare this to `return:` which serves a comparable function to `/local`, and which does not have this problem. Still, the design choice has been made to have a `refinement!` value for the "local variable indicator". Therefore I think that `/local` as injected by `has` and `function` should be bound to the same context as when they are specified explicitly by the user (i.e. not "injected"), thus the global context.

**Comment by greggirwin** [comment]
Excellent point on `return:` @meijeru. However, `/local` is not global when used manually.
```
>> get first spec-of func [/local x][]
*** Script Error: local has no value
*** Where: get
*** Stack:  

>> get first spec-of func [return: [string!]][]
== make native! [["Returns a value from a function" 
    value [any-type!]
]]
```

As a note, refinements are part of `all-word!`, not `any-word!`, making them more like issues. e.g. you can't `set` a refinement.

**Comment by meijeru** [comment]
It is difficult to determine the context of a refinement, since `context?` does not admit a `refinement!` value as argument. If you do `context? to-word first spec-of...` you get the global context.

**Comment by greggirwin** [comment]
True, but you have to use `to word!` because `context?` also doesn't support refinements.

Let me refine my statement: Do _refinements_ and _issues_ (and potentially other non-any-word! types) have to be bound to a context, or can they not be bound and considered only literal data?

**Comment by hiiamboris** [comment]
Keep it simple approach tells me that they have to. To not spawn special cases.

**Comment by 9214** [comment]
`any-word!` contains values that have bindings, `all-word!` adds on top of that values that do not explicitly have them.
```red
>> exclude all-word! any-word!
== make typeset! [refinement! issue!]
```
In practice this is not followed strictly, since some natives accept refinements as if they are words referring to values.
```red
>> foo: 'bar
== bar
>> get /foo
== bar
```
Internally both `issue!` and `refinement!` have a binding; at least `issue!`s are always bound to the global context.
```red
>> find mold words-of system/words "abracadabra"
== none
>> #abracadabra
== #abracadabra
>> find mold words-of system/words "abracadabra"
== "abracadabra]"
```

**Comment by hiiamboris** [comment]
Yes. And let's not forget that `to word! to refinement! word` should be lossless.

**Comment by dockimbel** [comment]
Refinements and issues are strictly symbolic, they should not be bound. `get` should error out on them.

> And let's not forget that to word! to refinement! word should be lossless.

No, refinements and words have different syntactic rules. The starting `/` can be followed by any valid word character (including digits).

**Comment by hiiamboris** [comment]
> No, refinements and words have different syntactic rules. The starting `/` can be followed by any valid word character (including digits).

@dockimbel Which reminds me of the unresolved design issue https://github.com/red/red/issues/650

**Comment by 9214** [comment]
@dockimbel I've updated `get` in https://github.com/red/red/pull/4594/ and resolved merge conflicts. It's ready to go.

**Comment by dockimbel** [comment]
`refinement!` has been removed from possible `get` argument types, so that the crash is now gone and an error is properly reported:
```
>> get first spec-of has [foo][]
*** Script Error: get does not allow refinement! for its word argument
*** Where: get
*** Stack:  

>> get first spec-of function [][foo:]
*** Script Error: get does not allow refinement! for its word argument
*** Where: get
*** Stack:  
```

**Comment by 9214** [comment]
@dockimbel I've made the same change in the above-mentioned PR 3 days ago and updated the respective tests.

**Comment by dockimbel** [comment]
The PR is making other unnecessary changes on refinements `ctx` field.
